Rewrite the `SuperchainERC20` page (#1273)
ethereum-optimism/docs: pages/stack/interop/superchain-erc20.mdx

Additional features:

*   **Simplified deployments**: Zero infrastructure cost to make your token cross-chain.
    Provides a consistent, unified implementation for tokens across all Superchain-compatible networks and a common cross-chain interface for the EVM ecosystem at large.
*   **Common standard**: Implements ERC-7802, a unified interface that can be used across all of Ethereum to enable cross-chain mint/burn functionality.

## How it works

[`SuperchainERC20`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainERC20.sol) and [`SuperchainTokenBridge`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainTokenBridge.sol) work together to allow ERC-20 tokens to be transferred from one chain to the other.

The initiating message burns tokens on the source chain.
The executing message then mints them on the destination chain.

```mermaid
sequenceDiagram
  box rgba(0,0,0,0.1) Source Chain

>     participant src-erc20 as ERC20<sup>1</sup>  

Reviewer: Can we double check the visual re <sup>

    participant src-bridge as Bridge
  end
  actor user as User  
  box rgba(0,0,0,0.1) Destination Chain

>     participant dst-l2Xl2 as L2ToL2

Reviewer: can we use the exact contract names so its easier for developers to grok the e2e flow

Author: I tried that first, but it resulted in tiny, unreadable letters. I'll try a different solution.

    participant dst-bridge as Bridge    
    participant dst-erc20 as ERC20 
  end
  rect rgba(0,0,0,0.1)
    note over src-erc20, dst-l2Xl2: Initiating Message
    user->>src-bridge: 1. Move n tokens
    src-bridge->>src-erc20: 2. Burn n tokens
    src-bridge->>dst-l2Xl2: 3. Relay n tokens to user
  end
  rect rgba(0,0,0,0.1)
    note over user,dst-erc20: Executing message
    user->>dst-l2Xl2: 4. Relay the message
    dst-l2Xl2->>dst-bridge: 4. Relay the message
    dst-bridge->>dst-erc20: 5. Mint n tokens to user
  end
```

> #### Initiating message

Reviewer: suggestion: add in paranthesis (source chain)

1.  The user (or a contract) calls [`SuperchainTokenBridge.sendERC20`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainTokenBridge.sol#L52-L78).

2.  The token bridge calls [`SuperchainERC20.crosschainBurn`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainERC20.sol#L37-L46) to burn those tokens on the source chain.

3.  The source token bridge calls [`SuperchainTokenBridge.relayERC20`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainTokenBridge.sol#L80-L97) on the destination token bridge.
    This call is relayed using [`L2ToL2CrossDomainMessenger`](./message-passing).

> #### Executing message

Reviewer: suggestion: add in paranthesis (destination chain)

> 4.  The user sends an executing message to [`L2ToL2CrossDomainMessenger`](./message-passing) to relay the message.

Reviewer: The user or relayer sends...

    <Callout type="warning">
      To do this securely you must either write the deployer in such a way that only a specific trusted ERC-20 contract, such as `SuperchainERC20`, can be deployed through it,
      or call the contract with `CREATE2` directly from an EOA you control.
      This is critical because if somebody can deploy a different ERC-20 contract at the same address, on any superchain network, they can mint themselves as many tokens as they'd like and then bridge them to the network where the original ERC-20 contract is located.
    </Callout>

{/* 

Add this after the tutorial is written

For step-by-step information on implementing SuperchainERC20, see [Deploy assets using SuperchainERC20](/stack/interop/assets/deploy-superchain-erc20)

*/}

## Comparison to other token implementations

`SuperchainERC20` differs from other token implementations in its focus and implementation:

*   `SuperchainERC20` implements ERC-7802, which provides a minimal crosschain mint/burn interface designed to be a common pattern for the EVM ecosystem.
*   `SuperchainERC20` uses the shared trust assumptions of the Superchain.

>     Traffic originating in *any* chain is trusted, because they are all managed by The Optimism Foundation.

Reviewer: managed by The OP FND -> share the same security standard (link to here: https://docs.optimism.io/superchain/standard-configuration)

Reviewer: what does "Traffic originating in *any* chain is trusted" mean?

I think it means Crosschain mint burns are facilitated for chains within the Superchain interop cluster bc of the same security standard?

import { Callout } from 'nextra/components'

import { InteropCallout } from '@/components/WipCallout'

<InteropCallout />

# SuperchainERC20

[`SuperchainERC20`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainERC20.sol) is an implementation of [ERC-7802](https://ethereum-magicians.org/t/erc-7802-crosschain-token-interface/21508) designed to enable asset interoperability in the Superchain.

> Asset interoperability allows for tokens to securely move across chains without asset wrapping or liquidity pools for maximum capital efficiency, thus unifying liquidity and simplifying the user experience.

Reviewer: Asset interoperability allows for tokens to move securely between chains via native minting and burning.  Asset interoperability solves the issues of liquidity fragmentation and poor user experiences caused by asset wrapping or liquidity pools and assets can essentially teleport from one blockchain to another, providing users with a secure and capital-efficient way to transact on the Superchain.

Reviewer: think this was closed by accident @qbzzt ?

Author: No, the latest version has the changes you asked for. See https://github.com/qbzzt/optimism-docs/blob/250122-erc20/pages/stack/interop/superchain-erc20.mdx#superchainerc20

3.  The source token bridge calls [`SuperchainTokenBridge.relayERC20`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainTokenBridge.sol#L80-L97) on the destination token bridge.
    This call is relayed using [`L2ToL2CrossDomainMessenger`](./message-passing).

#### Executing message (destination chain)

4.  The user, or a relayer, sends an executing message to [`L2ToL2CrossDomainMessenger`](./message-passing) to relay the message.

5.  The destination token bridge calls [`SuperchainERC20.crosschainMint`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainERC20.sol#L26-L35) to mint tokens for the user/contract that called `SuperchainTokenBridge.sendERC20` originally.

## Requirements

Application developers must do two things to make their tokens `SuperchainERC20` compatible.
Doing this setup now ensures that tokens benefit from interop (once it is available).

*   Grant permission to `SuperchainTokenBridge` (address `0x4200000000000000000000000000000000000028`) to call `crosschainMint` and `crosschainBurn`.

>     [`SuperchainERC20`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainERC20.sol) already does this for you.

Reviewer: If you are using SuperchianERC20 this is already done for you (add a comment for a future guide:, if not, here are the steps to do it )

Reviewer: was this meant to be closed @qbzzt ?

Author: Yes. I added the language you asked for.

import { Callout } from 'nextra/components'

import { InteropCallout } from '@/components/WipCallout'

<InteropCallout />

# SuperchainERC20

> [`SuperchainERC20`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainERC20.sol) is an implementation of [ERC-7802](https://ethereum-magicians.org/t/erc-7802-crosschain-token-interface/21508) designed to enable asset interoperability in the Superchain.
> Asset interoperability allows tokens to move securely in the Superchain by burning tokens on the source chain and minting the same number of tokens that were burned on the destination chain.
> Asset interoperability solves the issues of liquidity fragmentation and poor user experiences caused by asset wrapping or liquidity pools.
> Instead, assets essentially teleport from one chain in the Superchain to another, providing users with a secure and capital-efficient way to transact within the Superchain.

Reviewer 2: ```suggestion
The [`SuperchainERC20`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainERC20.sol) contract implements [ERC-7802](https://ethereum-magicians.org/t/erc-7802-crosschain-token-interface/21508) to enable asset interoperability within the Superchain. 

Asset interoperability allows tokens to move securely across the Superchain by burning tokens on the source chain and minting an equivalent amount on the destination chain. This approach addresses issues such as liquidity fragmentation and poor user experiences caused by asset wrapping or reliance on liquidity pools.

Instead of wrapping assets, this mechanism effectively "teleports" tokens between chains in the Superchain. It provides users with a secure and capital-efficient method for transacting across chains.

```

2.  The token bridge calls [`SuperchainERC20.crosschainBurn`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainERC20.sol#L37-L46) to burn those tokens on the source chain.

3.  The source token bridge calls [`SuperchainTokenBridge.relayERC20`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainTokenBridge.sol#L80-L97) on the destination token bridge.
    This call is relayed using [`L2ToL2CrossDomainMessenger`](./message-passing).

#### Executing message (destination chain)

4.  The user, or a relayer, sends an executing message to [`L2ToL2CrossDomainMessenger`](./message-passing) to relay the message.

5.  The destination token bridge calls [`SuperchainERC20.crosschainMint`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainERC20.sol#L26-L35) to mint tokens for the user/contract that called `SuperchainTokenBridge.sendERC20` originally.

## Requirements

> Application developers must do two things to make their tokens `SuperchainERC20` compatible.
> Doing this setup now ensures that tokens benefit from interop (once it is available).

Reviewer 2: ```suggestion
Application developers must complete two steps to make their tokens compatible with `SuperchainERC20`. 
Setting this up in advance ensures tokens will benefit from interop when it becomes available.
```

*   Grant permission to `SuperchainTokenBridge` (address `0x4200000000000000000000000000000000000028`) to call `crosschainMint` and `crosschainBurn`.
    If you are using [`SuperchainERC20`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainERC20.sol) this is already done for you.

      {/* 

      Add this after the tutorial is written

>       For step-by-step information on implementing SuperchainERC20, see [Deploy assets using SuperchainERC20](/stack/interop/assets/deploy-superchain-erc20)

Reviewer 2: ```suggestion
For detailed, step-by-step instructions on implementing SuperchainERC20, refer to [Deploy assets using SuperchainERC20](/stack/interop/assets/deploy-superchain-erc20).
```

      */}

*   Deploy the ERC-20 contract at the same address on every chain in the Superchain.
    This is easiest when using [`create2`](https://book.getfoundry.sh/tutorials/create2-tutorial).

>     <Callout type="warning">
>       To do this securely you must either write the deployer in such a way that only a specific trusted ERC-20 contract, such as `SuperchainERC20`, can be deployed through it,
>       or call the contract with `CREATE2` directly from an EOA you control.
>       This is critical because if somebody can deploy a different ERC-20 contract at the same address, on any superchain network, they can mint themselves as many tokens as they'd like and then bridge them to the network where the original ERC-20 contract is located.
>     </Callout>

Reviewer 2: ```suggestion
<Callout type="warning">
To ensure security, you must either design the deployer to allow only a specific trusted ERC-20 contract, such as `SuperchainERC20`, to be deployed through it, or call `CREATE2` to deploy the contract directly from an EOA you control. 

This precaution is critical because if an unauthorized ERC-20 contract is deployed at the same address on any Superchain network, it could allow malicious actors to mint unlimited tokens and bridge them to the network where the original ERC-20 contract resides.
</Callout>

```

## Comparison to other token implementations

> `SuperchainERC20` differs from other token implementations in its focus and implementation:
> 
> *   `SuperchainERC20` implements ERC-7802, which provides a minimal crosschain mint/burn interface designed to be a common pattern for the EVM ecosystem.
> *   `SuperchainERC20` uses the shared trust assumptions of the Superchain.
>     Traffic originating in any of the Superchain's chains is trusted, because all those chains [share the same security standard](https://docs.optimism.io/superchain/standard-configuration).

Reviewer 2: ```suggestion
`SuperchainERC20` differs from other token implementations in both focus and design:

*   It implements `ERC-7802`, a minimal cross-chain mint/burn interface designed to establish a common standard across the EVM ecosystem.  
*   It relies on the shared trust assumptions of the Superchain. All traffic originating from any chain within the Superchain is trusted because these chains [share the same security standard](/superchain/standard-configuration).

```

<Callout type="info">
  Projects moving from other token implementations may need to adapt to the `SuperchainERC20` specification.
</Callout>

## FAQ

### What happens if I bridge to a chain that does not have the ERC-20 contract?

> The initiating message will successfully burn the tokens on the source chain.
> However, the executing message will try to call `crosschainMint` on a non-existent contract and fail.
> Once a `SuperchainERC20` contract is properly deployed at the destination chain, you can attempt the executing message again and get your tokens.
> 

Reviewer 2: ```suggestion
The initiating message will successfully burn the tokens on the source chain. 
However, the executing message will fail because it attempts to call `crosschainMint` on a non-existent contract. 
Once a `SuperchainERC20` contract is properly deployed on the destination chain, you can retry the executing message to retrieve your tokens.
```
